### PR TITLE
Voting

### PR DESCRIPTION
**Pull request by zengqiangHe**
Hi there,

with this pull request I have:

- created a voting page which the user can vote "Bock" or "Wurst" for the upcoming event.
- connected the database and sent the vote results to there successfully.
- built a text on the event card, which can show the vote results.

Thanks for reviewing my code!

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -8,11 +8,13 @@ import { Route, Routes, useNavigate } from 'react-router-dom';
 import HomePageTabs from './components/homePageTabs/HomePageTabs';
 import EditEventPage from './pages/EditEventPage';
 import Navbar from './components/navbar/Navbar';
+import Voting from './pages/Voting';
 import NamePrompt from './pages/NamePrompt';
 
 function App() {
   const navigate = useNavigate();
   const [eventsList, setEventsList] = useState([]);
+  console.log(eventsList)
```
Comment (PReichetanz) [inline client/src/App.js:17]:
```suggestion
```

File: client/src/App.js [hunk 2]
```diff
@@ -40,13 +42,37 @@ function App() {
       .then((res) => {
         setEventsList([res, ...eventsList]);
         navigate('/');
+        console.log('saved new event');
+        console.log(res);
```
Comment (PReichetanz) [inline client/src/App.js:46]:
```suggestion
```

File: client/src/App.js [hunk 3]
```diff
@@ -40,13 +42,37 @@ function App() {
...
       })
       .catch((error) => {
         console.log('fail to add new events');
         console.log(error);
```
Comment (PReichetanz) [inline client/src/App.js:50]:
```suggestion
```

File: client/src/App.js [hunk 4]
```diff
@@ -40,13 +42,37 @@ function App() {
...
+        userName: name,
+      }),
+    })
+      .then((response) => response.json())
+      .then((res) => {
+        navigate('/');
+      })
+      .catch((error) => {
+        console.log('fail to add vote');
+        console.log(error);
```
Comment (PReichetanz) [inline client/src/App.js:72]:
```suggestion
```

File: client/src/components/eventCard/EventCard.jsx [hunk 5]
```diff
@@ -5,8 +5,23 @@ import delete_icon from '../../assets/delete_icon.png';
...
 import calendar_icon from '../../assets/calendar_icon.png';
+import { useState, useEffect } from 'react';
 
 const EventCard = ({ showEditButton, eventDetail, deleteEvent }) => {
+  const [confirmedText, setConfirmedText] = useState(' ');
+  useEffect(() => {
+    if (eventDetail.votes.length === 1 || eventDetail.votes.length === 0)
+      setConfirmedText('Niemand hat bisher Bock.');
+    else if (eventDetail.votes.length === 2)
+      setConfirmedText(eventDetail.votes[1].id + ' hat Bock.');
```
Comment (PReichetanz) [inline client/src/components/eventCard/EventCard.jsx:16]: I assumed that `eventDetail.votes[1].id` is an id (a number or a string), but not a name.

File: client/src/pages/Voting.jsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import location_icon from '../assets/location_icon.png';
+import clock_icon from '../assets/clock_icon.png';
+import calendar_icon from '../assets/calendar_icon.png';
+import styled from 'styled-components';
+import { useParams } from 'react-router-dom';
+
+export default function Voting({ events, setVotingConfirmation }) {
+  const { _id } = useParams();
+
+  const handleOnclickBockButton = (event) => {
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:11]:
```suggestion
  const handleOnClickBockButton = (event) => {
```

File: client/src/pages/Voting.jsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    event.preventDefault();
+    setVotingConfirmation(_id, true);
+  };
+
+  const handleOnclickWurstButton = (event) => {
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:16]:
```suggestion
  const handleOnClickWurstButton = (event) => {
```

File: client/src/pages/Voting.jsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    event.preventDefault();
+    setVotingConfirmation(_id, false);
+  };
+
+  const event = events.find((event) => event._id === _id);
+  if (event === undefined) {
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:22]: You might also write

```suggestion
  if (!event) {
```

File: client/src/pages/Voting.jsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+  return (
+    <>
+      <CardWrapper>
+        <h2>{event.title}</h2>
+        <p>{event.text}</p>
+        <p>
+          <img src={calendar_icon} alt="calendar icon" width="15" />
+          {event.date}
+        </p>
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:34]: Why do you wrap the img element inside a p element? This is not semantical; if it is about the block vs. inline style, you can use `display: block` (or inline-block) for the img.

File: client/src/pages/Voting.jsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        <p>
+          <img src={clock_icon} alt="time icon" width="15" />
+          {event.time}
+        </p>
+        <p>
+          <img src={location_icon} alt="location icon" width="15" />
+          {event.location}
+        </p>
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:42]: Same here.

File: client/src/pages/Voting.jsx [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+      </CardWrapper>
+      <ButtonWrapper>
+        <BockButton onClick={handleOnclickBockButton}>Bock</BockButton>
+        <WurstButton onClick={handleOnclickWurstButton}>Wurst</WurstButton>
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:46]:
```suggestion
        <BockButton onClick={handleOnClickBockButton}>Bock</BockButton>
        <WurstButton onClick={handleOnClickWurstButton}>Wurst</WurstButton>
```

File: client/src/pages/Voting.jsx [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      </ButtonWrapper>
+    </>
+  );
+}
+const CardWrapper = styled.li`
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:51]: Where exactly is your corresponding list as parent element? Wouldn't it make more sense to have a section or something similar here?

File: client/src/pages/Voting.jsx [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  background-color: purple;
+  border-radius: 2rem;
+  padding: 1.5rem;
+  margin: 0.5rem;
+  &:hover {
+    background-color: pink;
+  }
+`;
+
+const ButtonWrapper = styled.li`
```
Comment (PReichetanz) [inline client/src/pages/Voting.jsx:66]: I am wondering why a wrapper with two buttons as children is a list item. Why do you use a list (item) for buttons at all?

File: server/routes/api.js [hunk 14]
```diff
@@ -54,4 +54,32 @@ router.post('/', (req, res, next) => {
...
 
+router.patch('/vote/:id', (req, res, next) => {
+  const id = req.params.id;
+  const isConfirmed = req.body.isConfirmed;
+  const userName = req.body.userName;
+
+  BockWurst.findById(id)
+    .then((data) => {
+      const votingEntry = data.votes.find((entry) => userName === entry.id);
+      if (votingEntry === undefined) {
```
Comment (PReichetanz) [inline server/routes/api.js:65]: You can also try
```suggestion
      if (!votingEntry) {
```

File: server/routes/api.js [hunk 15]
```diff
@@ -54,4 +54,32 @@ router.post('/', (req, res, next) => {
...
+      data
+        .save()
+        .then((data) => {
+          res.status(201).send(data);
+        })
+        .catch((error) => {
+          res.status(400).json({ error: error.message });
+        });
+    })
+
```
Comment (PReichetanz) [inline server/routes/api.js:79]:
```suggestion
```

**Review by PReichetanz — CHANGES_REQUESTED** [review]
Looks good so far, I add some issues to consider.

File: client/src/App.js (hunk 1 above)
Comment (zengqiangHe) [reply client/src/App.js:17]: Done!

File: client/src/App.js (hunk 3 above)
Comment (zengqiangHe) [reply client/src/App.js:50]: Done!

File: client/src/App.js (hunk 4 above)
Comment (zengqiangHe) [reply client/src/App.js:72]: Done!

File: client/src/components/eventCard/EventCard.jsx (hunk 5 above)
Comment (zengqiangHe) [reply client/src/components/eventCard/EventCard.jsx:16]: You're absolutely right! I have to change it in my database.

File: client/src/pages/Voting.jsx (hunk 6 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:11]: What is the problem here? I cannot see the difference.

File: client/src/pages/Voting.jsx (hunk 7 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:16]: What is the problem here? I cannot see the difference.

File: client/src/pages/Voting.jsx (hunk 8 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:22]: Done

File: client/src/pages/Voting.jsx (hunk 9 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:34]: Done, thanks.

File: client/src/pages/Voting.jsx (hunk 10 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:42]: Done.

File: client/src/pages/Voting.jsx (hunk 11 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:46]: What is the problem here? I cannot see the difference.

File: client/src/pages/Voting.jsx (hunk 12 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:51]: Done, thanks!

File: client/src/pages/Voting.jsx (hunk 13 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:66]: I just want to use flexbox giving the two buttons some space in between, so I didn't really consider why I used li (item). Is it better to use div? Or button? I am not quite sure.

File: server/routes/api.js (hunk 14 above)
Comment (zengqiangHe) [reply server/routes/api.js:65]: Done, thanks!

File: client/src/pages/Voting.jsx (hunk 11 above)
Comment (PReichetanz) [reply client/src/pages/Voting.jsx:46]: I just changed the lower case `click` to upper case `Click`, but it is quite subjective I guess ;)

File: client/src/pages/Voting.jsx (hunk 7 above)
Comment (PReichetanz) [reply client/src/pages/Voting.jsx:16]: See my comment below.

File: client/src/pages/Voting.jsx (hunk 6 above)
Comment (PReichetanz) [reply client/src/pages/Voting.jsx:11]: See my comment below.

File: client/src/pages/Voting.jsx (hunk 11 above)
Comment (zengqiangHe) [reply client/src/pages/Voting.jsx:46]: Oh, I got it!

**Pull request merged**